api/v1alpha1/extension_types.go

```diff
@@ -0,0 +1,169 @@
+	//+kubebuilder:Optional
+	// Version is an optional semver constraint on the package version. If not specified, the latest version available of the package will be installed.
+	// If specified, the specific version of the package will be installed so long as it is available in any of the content sources available.
+	// Examples: 1.2.3, 1.0.0-alpha, 1.0.0-rc.1
+	//
+	// For more information on semver, please see https://semver.org/
+	// version constraint definition
+	Version string `json:"version,omitempty"`
+
+	//+kubebuilder:validation:MaxLength:=48
+	//+kubebuilder:validation:Pattern:=^[a-z0-9]+([\.-][a-z0-9]+)*$
+	// channel constraint definition
+	Channel string `json:"channel,omitempty"`
+
+	//+kubebuilder:validation:Enum:=Enforce;Ignore
+	//+kubebuilder:default:=Enforce
+	//+kubebuilder:Optional
+	//
+	// upgradeConstraintPolicy Defines the policy for how to handle upgrade constraints
+	UpgradeConstraintPolicy UpgradeConstraintPolicy `json:"upgradeConstraintPolicy,omitempty"`
+}
+
+// TODO: Implement ExtensionSourceDirect containing a URL or other reference mechanism
+
+type ExtensionSource struct {
+	// A source package defined by a name, version and/or channel
+	Package *ExtensionSourcePackage `json:"package,omitempty"`
+
+	// A direct package defined by a URL
+	// TODO: Direct *ExtensionSourceDirect `json:"direct,omitempty"`
+}
+
+// ExtensionSpec defines the desired state of Extension
+type ExtensionSpec struct {
+	//+kubebuilder:validation:Enum:=Active;Paused
+	//+kubebuilder:default:=Active
+	//+kubebuilder:Optional
+	//
+	// managed controls the management state of the extension. "Active" means this extension will be reconciled and "Paused" means this extension will be ignored.
+	Managed ExtensionManagedState `json:"managed,omitempty"`
+
+	//+kubebuilder:validation:MaxLength:=253
+	//+kubebuilder:validation:Pattern:=^[a-z0-9]+([\.-][a-z0-9]+)*$
+	//
+	// serviceAccountName is he name of a service account in the Extension's namespace that will be used to manage the installation and lifecycle of the extension.
+	ServiceAccountName string `json:"serviceAccountName"`
+
+	//+kubebuilder:validation:MaxLength:=253
+	//+kubebuilder:validation:Pattern:=^[a-z0-9]+([\.-][a-z0-9]+)*$
+	//+kubebuilder:Optional
+	//
+	// defaultNamespace is the location of installation if different than the resource namespace
+	DefaultNamespace string `json:"defaultNamespace,omitempty"`
+
+	// source of Extension to be installed
+	Source ExtensionSource `json:"source"`
+
+	//+kubebuilder:Optional
+	//
+	// watchNamespaces indicates which namespaces the extension should watch.
+	// This feature is currently supported only with RegistryV1 bundles.
+	WatchNamespaces []string `json:"watchNamespaces,omitempty"`
```

Comment:
Let's drop `WatchNamespace` for now too. This is what I was proposing in our OLM WG sync this morning. I think we can add in a values field later based on this issue https://github.com/operator-framework/operator-controller/issues/612.

Reply:
This is currently supported by rukpak, but I ought to be able to get rid of it, and remove any appropriate tests.